src/core/stream_set.c

```diff
@@ -303,50 +368,59 @@ QuicStreamSetInitializeTransportParameters(
-        CxPlatHashtableEnumerateBegin(StreamSet->StreamTable, &Enumerator);
-        while ((Entry = CxPlatHashtableEnumerateNext(StreamSet->StreamTable, &Enumerator)) != NULL) {
-            QUIC_STREAM* Stream = CXPLAT_CONTAINING_RECORD(Entry, QUIC_STREAM, TableEntry);
+    CXPLAT_LIST_ENTRY* Link = StreamSet->WaitingStreams.Flink;
+    while (Link != &StreamSet->WaitingStreams) {
+        QUIC_STREAM* Stream =
+            CXPLAT_CONTAINING_RECORD(Link, QUIC_STREAM, WaitingLink);
+        Link = Link->Flink;
 
-            uint8_t FlowBlockedFlagsToRemove = 0;
+        const uint64_t StreamType = Stream->ID & STREAM_ID_MASK;
+        const uint64_t StreamIndex = (Stream->ID >> 2);
+        const QUIC_STREAM_TYPE_INFO* Info = &Stream->Connection->Streams.Types[StreamType];
 
-            uint64_t StreamType = Stream->ID & STREAM_ID_MASK;
-            uint64_t StreamCount = (Stream->ID >> 2) + 1;
-            const QUIC_STREAM_TYPE_INFO* Info =
-                &Stream->Connection->Streams.Types[StreamType];
-            if (Info->MaxTotalStreamCount >= StreamCount &&
-                Stream->OutFlowBlockedReasons & QUIC_FLOW_BLOCKED_STREAM_ID_FLOW_CONTROL) {
-                FlowBlockedFlagsToRemove |= QUIC_FLOW_BLOCKED_STREAM_ID_FLOW_CONTROL;
-                QuicStreamIndicatePeerAccepted(Stream);
-            } else {
-                QuicSendSetSendFlag(
-                    &Stream->Connection->Send,
-                    STREAM_ID_IS_UNI_DIR(Stream->ID) ?
-                        QUIC_CONN_SEND_FLAG_UNI_STREAMS_BLOCKED : QUIC_CONN_SEND_FLAG_BIDI_STREAMS_BLOCKED);
-            }
+        CXPLAT_DBG_ASSERT(Stream->OutFlowBlockedReasons & QUIC_FLOW_BLOCKED_STREAM_ID_FLOW_CONTROL);
 
-            uint64_t NewMaxAllowedSendOffset =
-                QuicStreamGetInitialMaxDataFromTP(
-                    Stream->ID,
-                    QuicConnIsServer(Connection),
-                    &Connection->PeerTransportParams);
+        uint8_t FlowBlockedFlagsToRemove = 0;
+        if (StreamIndex < Info->MaxTotalStreamCount) {
+            FlowBlockedFlagsToRemove |= QUIC_FLOW_BLOCKED_STREAM_ID_FLOW_CONTROL;
+            CxPlatListEntryRemove(&Stream->WaitingLink);
+            Stream->Flags.InWaitingList = FALSE;
 
-            if (Stream->MaxAllowedSendOffset < NewMaxAllowedSendOffset) {
-                Stream->MaxAllowedSendOffset = NewMaxAllowedSendOffset;
-                FlowBlockedFlagsToRemove |= QUIC_FLOW_BLOCKED_STREAM_FLOW_CONTROL;
-                Stream->SendWindow = (uint32_t)CXPLAT_MIN(Stream->MaxAllowedSendOffset, UINT32_MAX);
+            CXPLAT_DBG_ASSERT(StreamSet->StreamTable != NULL);
+            if (!QuicStreamSetInsertStream(StreamSet, Stream)) {
+                //
+                // The stream hash-table should have been initialized
+                // already when inserting stream in `WaitingStreams`
+                //
+                CXPLAT_FRE_ASSERTMSG(FALSE, "Steam table lazy intialization failed");
             }
```

Comment:
```suggestion
            //
            // The stream hash-table should have been initialized
            // already when inserting stream in `WaitingStreams`
            //
            CXPLAT_FRE_ASSERTMSG(
                QuicStreamSetInsertStream(StreamSet, Stream),
                "Steam table lazy intialization failed");
```

Reply:
I generally prefer avoiding code with side effect in asserts, I find it error prone. For DBG assert, the call is skipped in release build, and FRE asserts look the same, so it always raises a warning flag when reading this code, "wait, does this actually execute?".
I can do the change if side effect in assert is a common practice in the codebase but would rather not otherwise.

Reply:
We do have a lot of code (see posix platform) that follows this pattern.